kong/plugins/opentelemetry/handler.lua

```diff
@@ -0,0 +1,185 @@
+  local spans_n = #spans
+  if spans_n == 0 then
+    return
+  end
+
+  local start = ngx_now()
+
+  -- cache http headers
+  local headers = default_headers
+  if conf.http_headers then
+    headers = http_headers_cache[conf.http_headers]
+  end
+
+  if not headers then
+    headers = clone(default_headers)
+    if conf.http_headers and conf.http_headers ~= null then
+      for k, v in pairs(conf.http_headers) do
+        headers[k] = v and v[1]
+      end
+    end
+    http_headers_cache[conf.http_headers] = headers
+  end
+
+  -- batch send spans
+  local batch_spans = tablepool_fetch(POOL_BATCH_SPANS, conf.batch_span_count, 0)
+
+  for i = 1, spans_cache.n do
+    local len = (batch_spans.n or 0) + 1
+    batch_spans[len] = spans_cache[i]
+    batch_spans.n = len
+
+    if len >= conf.batch_span_count then
+      local pb_data = encode_traces(batch_spans, conf.resource_attributes)
+      clear(batch_spans)
+
+      http_export_request(conf, pb_data, headers)
+    end
+  end
+
+  -- remain spans
+  if spans_cache.n and spans_cache.n > 0 then
+    local pb_data = encode_traces(batch_spans, conf.resource_attributes)
+    http_export_request(conf, pb_data, headers)
+  end
+
+  clear(spans_cache)
+
+  tablepool_release(POOL_BATCH_SPANS, batch_spans)
+
+  ngx.update_time()
+  local duration = ngx.now() - start
+  ngx_log(ngx_DEBUG, _log_prefix, "opentelemetry exporter sent " .. spans_n ..
+    " traces to " .. conf.http_endpoint .. " in " .. duration .. " seconds")
+end
+
+local function process_span(span)
+  if span.should_sample == false
+      or kong.ctx.plugin.should_sample == false
+  then
+    -- ignore
+    return
+  end
+
+  -- overwrite
+  local trace_id = kong.ctx.plugin.trace_id
+  if trace_id then
+    span.trace_id = trace_id
+  end
+
+  local pb_span = transform_span(span)
+
+  local len = spans_cache.n or 0
+  len = len + 1
+
+  spans_cache[len] = pb_span
+  spans_cache.n = len
+end
+
+function OpenTelemetryHandler:rewrite()
+  local headers = ngx_get_headers()
+  local root_span = ngx.ctx.KONG_SPANS and ngx.ctx.KONG_SPANS[1]
+  -- tracing instrumetation not enabled
+  if not root_span then
+    local tracer = kong.tracing.new("otel")
+    root_span = tracer.start_span("otel")
+    kong.ctx.plugin.should_sample = false
+  end
+
+  local header_type, trace_id, span_id, _, should_sample, _ = propagation_parse(headers)
+  if should_sample == false then
+    root_span.should_sample = should_sample
+  end
+
+  -- overwrite trace id
+  if trace_id then
+    root_span.trace_id = trace_id
+    kong.ctx.plugin.trace_id = trace_id
+  end
+
+  if span_id then
+    root_span.parent_id = span_id
+  end
+
+  propagation_set("w3c", header_type, root_span)
+end
+
+function OpenTelemetryHandler:log(conf)
+  ngx_log(ngx_DEBUG, _log_prefix, "total spans in current request: ", ngx.ctx.KONG_SPANS and #ngx.ctx.KONG_SPANS)
+
+  -- transform spans
+  kong.tracing.process_span(process_span)
+
+  local last = last_run_cache[conf] or 0
```

Comment:
discussed offline, `conf` will be a new table everytime, we can use `conf.__key__` to distinguish different plugin instances.